central/xlsx/device.go

```
@@ -0,0 +1,174 @@
+						continue
+					}
+
+					// check if the mapping object is defined in the Devices sheet if the defaultValue is not empty
+					// if not, insert the mapping object as a new column in the Devices sheet with defaultValue set in each data row
+					if mapping.defaultValue != "" {
+						err = checkMappingObject(xlsFile, devicesSheetName, &colCount, len(rows), mapping.defaultValue, objectField, &header)
+						if err != nil {
+							return fmt.Errorf("failed to check mapping object: %w", err)
+						}
+					}
+				}
+			}
+		}
+		break
+	}
+
+	rows, err = xlsFile.GetRows(devicesSheetName)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve all rows from %s after inserting misshing columns: %w", devicesSheetName, err)
+	}
+
+	// parse the device data rows
+	for rowIndex, row := range rows {
+		if rowIndex == 0 {
+			continue
+		}
+
+		convertedDevice := dtos.Device{ProtocolName: protocol}
+		_, err = readStruct(&convertedDevice, protocol, header, row)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal an excel row into Device DTO: %w", err)
+		}
+
+		// validate the device DTO
+		err := common.Validate(convertedDevice)
+		if err != nil {
+			deviceErr := fmt.Errorf("device %s validation error: %v", convertedDevice.Name, err)
+			deviceXlsx.ValidateErrors = append(deviceXlsx.ValidateErrors, deviceErr)
+		} else {
+			deviceXlsx.Devices = append(deviceXlsx.Devices, &convertedDevice)
+		}
+	}
+
+	err = deviceXlsx.convertAutoEvents(xlsFile)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// convertAutoEvents parses the AutoEvents sheet and convert the rows to AutoEvent DTOs
+func (deviceXlsx *deviceXlsx) convertAutoEvents(xlsFile *excelize.File) error {
+	var header []string
+
+	rows, err := xlsFile.GetRows(autoEventsSheetName)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve all rows from %s worksheet: %w", autoEventsSheetName, err)
```

auto event is optional, and it can be ignored

---

Modified to ignore the error on the caller

---

Done, the case for no AutoEvents worksheet or no rows in it will be ignored.